rep-2015.rst

```diff
@@ -0,0 +1,137 @@
+    │           ├── cert.pem
+    │           ├── key.pem
+    │           ├── ...
+    └── public
+        ├── ...
+
+Note that it also requires the names of the files to be the ones expected.
+Specifically, all certificate and key files must have the ``.pem`` extension.
+In order to configure the security properties of a DDS participant, the path of the appropriate file in the enclave directory is added as the ``file`` URI of the corresponding property.
+For example, for the private key in the authentication plugin:
+
+.. code-block::
+
+    dds.sec.auth.private_key = file:<root>/enclaves/foo/bar/key.pem
+
+
+Terminology
+===========
+
+:Certificate:
+  A data file containing the identity and authentication information of an entity.
+
+:PKCS#11:
+  Refers to one of the Public-Key Cryptography Standards, defined in PKCS#11 Standard [2].
+
+:Keystore:
+  The keystore is the root directory where the DDS security artifacts are stored.
+  See ROS 2 Security Enclaves [3] for further information.
+
+:Enclave:
+  The process or group of processes that will share the same identity and access control rules.
+  See ROS 2 Security Enclaves [3] for further information.
+
+:URI:
+  A resource's identifier that specifies how it can be accessed.
+
+Specification
+=============
+
+Goals
+-----
+
+Support PKCS#11 URIs for certificates and key files.
+``data`` URIs are out of scope of this proposal.
+
+Proposal
+--------
+
+We want to keep the current keystore structure as much as possible, as this will enable to keep all the current implementations regarding the enclave management and the CLI features that help setting up the keystore.
+No changes should be needed to systems that do not use the PKCS#11 scheme.
+
+The problem then is how to let the RMW implementation when we want to use a ``file`` URI and when a ``pkcs11`` URI; and how to provide the values of these URIs (i.e., the file path in the case of ``file`` and the token name in the case of ``pkcs11``).
+
+This can be solved if we allow for the key and certificate files in the enclave to have ``.pem`` or ``.p11`` extensions.
+Files with ``.p11`` will contain the PKCS#11 URI instead of the actual document data.
+Then, the RMW can be aware of the file extension and set the security property accordingly.
+Taking the private key in the authentication plugin as an example:
+
+#. The RMW will look for a file with name ``key.pem`` or ``key.p11`` in the enclave.
+#. If there is a ``key.p11`` file, and the RMW supports ``pkcs11`` URIs, it must load the content of the file, and set this content as the value of the property.
+   Some check can be done at this point, e.g., assert that the file contains a valid URI (i.e., starts with ``pkcs11:``).
+#. Otherwise, if there is a ``key.pem`` file, it keeps the current behavior, and sets the value of the property to the path of the file.
+
+With this proposal, current RMW implementations do not need to be updated if no support for PKCS#11 URIs is planned.
```

Comment:
Out of curiosity, which rmw implementation actually supports `pkcs11` scheme? especially tier I implementations. it would be probably nice to have all tier I implementations support this `pkcs11` scheme for consistent user application behavior.

Reply:
AFAIK, both Connext and Fast DDS support `pkcs11`. I'm not sure about Cyclone DDS